lib/trivial: add assertMsg and assertMsgOneOf

Profpatsch opened the pull request:
This is a cute little trick to make nix print an error message if an assertion fails.
It was lingering in my local repo for quite some time.

Profpatsch committed:
lib/trivial: add assertMsg

Infinisil commented on lib/trivial.nix:
```diff
@@ -188,6 +188,26 @@ rec {
+  assertMsg = pred: msg:
+    if pred
+    then true
+    else builtins.trace msg false;
```
Why not make the `assert` part of the function?

Profpatsch replied:
Hm, it has to do with the fact that `assert` is not really an expression (it doesn’t return a value) but a statement (that influences the state of evaluation) of sorts.

We could change it to include an assert, but that means it can’t appear in statement position any longer. Maybe that’s not bad, though.

Profpatsch replied:
```
let
  assertMsgIf = pred: msg: v: if pred v then v else builtins.trace msg; assert false; /*never reached*/ {};
  assertMsg = pred: assertMsgIf (const pred)
in
  (assertMsg (bar != borz)) /*where to put the derivation?*/

Infinisil replied:
Oh right, makes sense

infinisil reviewed (approved):
Looks good to me then

zimbatm reviewed (commented):
Are there places where it could be used in nixpkgs?

Profpatsch commented:
> Are there places where it could be used in nixpkgs?

Every time you want to throw a more helpful message for a failing assert. I did have some examples in code that was not pushed ultimately.

Profpatsch committed:
lib/trivial: add assertOneOf

Profpatsch committed:
lib/trivial: add a few examples of usage of assertMsg/assertOneOf

Profpatsch commented:
I’ve added a few examples for the asserts in `lib`, also another helper for enumerations called `assertOneOf`.

zimbatm reviewed (approved):
Looks good! It's the start of a good assertion-testing library.

Infinisil commented on lib/trivial.nix:
```diff
@@ -188,6 +188,44 @@ rec {
+    (lib.elem val xs)
+    "${name} must be one of ${
+      lib.generators.toPretty {} xs}, but is: ${
+        lib.generators.toPretty {} val}";
```
How about creating a more general `assertType`, which takes a type and calls `.check` on the value. Then this function is a specialization of it with `type = types.enum xs`.

edolstra commented:
IMHO, these functions should not be in `trivial.nix` since they are not, in fact, trivial. (For example, they call `lib.generators.toPretty` which is very non-trivial.)

zimbatm commented:
good point. if it's the start of a new assertion system it might make sense to group them in a new `lib/assets.nix` file

Profpatsch committed:
lib: move assertMsg and assertOneOf to their own library file

Since the `assertOneOf` uses `lib.generators`, they are not really trivial
anymore and should go into their own library file.

Profpatsch commented:
@edolstra @zimbatm done.

oxij commented:
- I think, syntactically, with current nix it would be nice to rename `assertMsg` to `withMessage` or something:

```
assert withMessage pred message
```

or

```
assert predWithMsg pred message
```

That function is `trivial`, btw. Should be reusable for other stuff like warnings too.

- I think `assertOneOf` should not exists, it does type checking for a specific type, I think we need to make `types-simple` to check the whole of NixOS, replace `types` with `types-simple` (I plan to get to that the next week), and then make a general type checking assert using those.

- Also, wouldn't it be nice if `assert` was a library function? We could then just define `assertWithMsg` too.

I think this should be pretty easy to do too. If we were to interpret top-level nix expression `a; b` as `seq a b` and define

```
assert = p: if p then null else throw "assert";
assertWithMsg = p: m: if p then null else throw m;
```

then `assert false; whatever` does exactly what you would expect. Migration paths is a bit tricky, though.

@edolstra, thoughts?

Profpatsch commented:
> rename `assertMsg` to `withMessage`

I thought about that and am open to changing the name(s).



> I think `assertOneOf` should not exists

Yeah, it’s pretty specific and should be generalized sometime, but I think it’s a good and pragmatic start. We can deprecate it or supersede it once we have figured out something more general.



> assert = p: if p then null else throw "assert";
> assertWithMsg = p: m: if p then null else throw m;

Can you create a minimal working example? I don’t see how that could work in practice.

oxij commented:
> name

I'm wholeheartedly for changing it.

> the specific thing

I'm like really ewww about this, stuff like that will start being used outside of `nixpkgs` and getting rid of it gracefully will take two whole releases.

>> assert = p: if p then null else throw "assert";
>> assertWithMsg = p: m: if p then null else throw m;
>
> Can you create a minimal working example? I don’t see how that could work in practice.

Like I said, you would need a patched nix for that. I mean nothing prevents nix from intepreting

```
assert false; "test"
```

more or less as

```
let
  nonBultinAssert = p: if p then null else throw "assert";
  assertWithMsg = p: m: if p then null else throw m;
in bultins.seq (nonBultinAssert false) "test"
```

Profpatsch commented:
Amended the last commit, apparently I missed a few unqualified uses after moving it out of `lib/trivial.nix`.

Profpatsch commented:
Any more comments? Otherwise I’ll merge in a bit.

Profpatsch commented:
@oxij You are free to generalize it, but I wouldn’t want to block this issue until somebody does.

kalbasit commented:
@Profpatsch @Infinisil can you cherry-pick this to release-18.09?

Profpatsch commented:
According to the IRC bot, it was backported (to somewhere at least, can’t tell).

kalbasit commented:
Hmm, I don't see the `lib/asserts.nix` file over on [release-18.09](https://github.com/NixOS/nixpkgs/tree/release-18.09/lib) branch.

Profpatsch commented on lib/asserts.nix:
```diff
@@ -0,0 +1,44 @@
+  assertMsg = pred: msg:
+    if pred
+    then true
+    else builtins.trace msg false;
```
Sounds like a good improvement to me.

Profpatsch replied:
Wait, how would that look in practice? Would it still be used like

```
assert (assertMsg ("foo" == "bar") "foo is not bar, silly"); ""
```

aszlig replied:
Sure, the difference is just that `assert` here essentially becomes a no-op if `pred` doesn't match, because `throw` already takes care of raising an error before the actual assertion fails.

Profpatsch replied:
Then the API stays the same, I see. I think we should do that.
